### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by faem**
- define interfaces, https://github.com/pharmer/pharmer/blob/refactor/cloud/api.go#L17-L45
each provider only needs to implement these interfaces, all common codes across providers are moved in `cloud/`

- remove `context` that was passed everywhere
before, we faced many nil pointer dereference errors due to contexts
most functions took context as parameters, it was unclear what values were stored in the context, what values were actually needed in function

- replace context with `scope` inspired by cluster-api-aws and azure implementations, things that were stored in context before are stored here. it gives a clear view of everything
https://github.com/pharmer/pharmer/blob/refactor/cloud/scope.go#L25-L32 

- minimize function parameters everywhere possible, before most functions took context as param, now functions takes only things they need as parameters

- improves testability. before that, we had to prepare context, load everything into the context and then pass it to that functions

- repackaging, before most utility functions were were dumped in `cloud/` library, we've tried to move these utility functions to separate packages

### CODE REVIEW

File: cloud/context.go [hunk 1]
```diff
@@ -150,11 +150,11 @@ func NewContext(parent context.Context, cfg *api.PharmerConfig, env _env.Environ
 	c = WithEnv(c, env)
 	c = WithNameGenerator(c, &api.NullNameGenerator{})
 	c = WithLogger(c, log.New(c))
-	c = WithStore(c, NewStoreProvider(parent, cfg))
+	c = WithStore(c, NewStoreProvider(cfg))
 	return c
 }
 
-func NewStoreProvider(ctx context.Context, cfg *api.PharmerConfig) store.Interface {
+func NewStoreProvider(cfg *api.PharmerConfig) store.Interface {
```
Comment (tahsinrahman) [inline cloud/context.go]: add comment

File: config/config.go [hunk 2]
```diff
@@ -90,10 +86,3 @@ func NewDefaultConfig() *api.PharmerConfig {
 		},
 	}
 }
-
-func NewStoreProvider(ctx context.Context, cfg *api.PharmerConfig) store.Interface {
```
Comment (tahsinrahman) [inline config/config.go]: this needs to be removed?

File: store/plugins.go [hunk 3]
```diff
@@ -12,7 +11,7 @@ import (
 // The config parameter provides an io.Reader handler to the factory in
```
Comment (tahsinrahman) [inline store/plugins.go]: fix comment

File: store/plugins.go [hunk 4]
```diff
@@ -58,12 +57,12 @@ func Providers() []string {
 // was known but failed to initialize. The config parameter specifies the
```
Comment (tahsinrahman) [inline store/plugins.go]: fix comment

File: cloud/create.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	// set cloud-specific configs
+	if err = cm.SetDefaultCluster(cluster); err != nil {
+		return nil, nil, errors.Wrap(err, "failed to set default cluster")
+	}
+
+	if !managedProviders.Has(cluster.ClusterConfig().Cloud.CloudProvider) {
+		for i := 0; i < cluster.Spec.Config.MasterCount; i++ {
+			master, err := CreateMasterMachines(cm, i)
```
Comment (tahsinrahman) [inline cloud/create.go]: Use name instead of int

File: cloud/create.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+func CreateMasterMachines(cm Interface, index int) (*clusterapi.Machine, error) {
+	cluster := cm.GetCluster()
+	providerSpec, err := cm.GetDefaultMachineProviderSpec(cluster, "", api.MasterRole)
+	if err != nil {
+		return nil, err
+	}
+
+	/*role := api.RoleMember
```
Comment (tahsinrahman) [inline cloud/create.go]: Remove

File: cloud/create.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	cluster := cm.GetCluster()
+
+	providerSpec, err := cm.GetDefaultMachineProviderSpec(cluster, sku, api.RoleNode)
+	if err != nil {
+		return errors.Wrap(err, "failed to get default machine provider spec")
+	}
+
+	machineSet := clusterapi.MachineSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              strings.Replace(strings.ToLower(sku), "_", "-", -1) + "-pool",
```
Comment (tahsinrahman) [inline cloud/create.go]: Common func

File: cloud/lib.go [hunk 8]
```diff
@@ -27,185 +26,132 @@ import (
...
-	config := cluster.Spec.Config
-	if cluster == nil {
-		return nil, errors.New("missing cluster")
-	} else if cluster.Name == "" {
-		return nil, errors.New("missing cluster name")
-	} else if config.KubernetesVersion == "" {
-		return nil, errors.New("missing cluster version")
-	}
+func getPharmerCerts(clusterName string) (*api.PharmerCertificates, error) {
+	pharmerCerts := &api.PharmerCertificates{}
```
Comment (tahsinrahman) [inline cloud/lib.go]: Use api consts

File: cloud/lib.go [hunk 9]
```diff
@@ -27,185 +26,132 @@ import (
...
-		for i := 0; i < cluster.Spec.Config.MasterCount; i++ {
-			master, err := CreateMasterMachines(ctx, cluster, i)
-			if err != nil {
-				return nil, err
-			}
-			if _, err = Store(ctx).Owner(owner).Machine(cluster.Name).Create(master); err != nil {
-				return nil, err
-			}
-		}
+	pubKey, privKey, err := LoadSSHKey(GenSSHKeyName(clusterName))
```
Comment (tahsinrahman) [inline cloud/lib.go]: Think again

File: Makefile [hunk 10]
```diff
@@ -272,6 +272,20 @@ release: docker-manifest
 	fi
 	@$(MAKE) clean all-push --no-print-directory
 
+.PHONY: install
```
Comment (tamalsaha) [inline Makefile]: We already have `make build` or just `make`. Why do we need this?

Comment (tamalsaha) [reply Makefile]:
```
.PHONY: install
install: build
    copy $(OUTBIN) ~/go/bin
```

File: apiserver/delete.go [hunk 11]
```diff
@@ -51,7 +51,7 @@ func (a *Apiserver) DeleteCluster() error {
 			noti := notification.NewNotifier(a.ctx, a.natsConn, strconv.Itoa(int(obj.ClusterID)))
 			newCtx := WithLogger(a.ctx, noti)
 
-			cluster, err = Delete(newCtx, cluster.Name, strconv.Itoa(int(obj.UserID)))
+			cluster, err = Delete(cluster.Name)
```
Comment (tahsinrahman) [inline apiserver/delete.go]: TODO

File: cloud/apiserver.go [hunk 12]
```diff
@@ -54,8 +56,9 @@ func IsNodeReady(node *core.Node) bool {
 	return false
 }
 
-func NewRestConfig(ctx context.Context, cluster *api.Cluster) (*rest.Config, error) {
-	adminCert, adminKey, err := CreateAdminCertificate(ctx)
+func NewRestConfig(cm Interface, cluster *api.Cluster) (*rest.Config, error) {
```
Comment (tahsinrahman) [inline cloud/apiserver.go]: pass only ca-certs instead of cm

File: cloud/apiserver.go [hunk 13]
```diff
@@ -79,31 +82,31 @@ func NewRestConfig(ctx context.Context, cluster *api.Cluster) (*rest.Config, err
 
 // WARNING:
 // Returned KubeClient uses admin client cert. This should only be used for cluster provisioning operations.
-func NewAdminClient(ctx context.Context, cluster *api.Cluster) (kubernetes.Interface, error) {
-	cfg, err := NewRestConfig(ctx, cluster)
+func NewAdminClient(cm Interface, cluster *api.Cluster) (kubernetes.Interface, error) {
```
Comment (tahsinrahman) [inline cloud/apiserver.go]: pass only ca-certs instead of cm

File: cloud/clusterapi.go [hunk 14]
```diff
@@ -22,27 +22,24 @@ import (
 )
 
 type ClusterApi struct {
-	ctx     context.Context
-	cluster *api.Cluster
-
-	namespace string
-	token     string
-	kc        kubernetes.Interface
+	ctx                 context.Context
```
Comment (tahsinrahman) [inline cloud/clusterapi.go]: pass only cm if possible

File: cloud/clusterapi.go [hunk 15]
```diff
@@ -186,7 +184,7 @@ func (ca *ClusterApi) CreateMachineController(controllerManager string) error {
 func (ca *ClusterApi) CreatePharmerSecret() error {
 	providerConfig := ca.cluster.ClusterConfig()
 
-	cred, err := Store(ca.ctx).Owner(ca.Owner).Credentials().Get(ca.cluster.ClusterConfig().CredentialName)
+	cred, err := Store(ca.ctx).Credentials().Get(ca.cluster.ClusterConfig().CredentialName)
```
Comment (tahsinrahman) [inline cloud/clusterapi.go]: remove store(ctx)

File: cloud/clusterapi.go [hunk 16]
```diff
@@ -221,7 +219,7 @@ func (ca *ClusterApi) CreatePharmerSecret() error {
 		return err
 	}
 
-	publicKey, privateKey, err := Store(ca.ctx).Owner(ca.Owner).SSHKeys(ca.cluster.Name).Get(ca.cluster.ClusterConfig().Cloud.SSHKeyName)
+	publicKey, privateKey, err := Store(ca.ctx).SSHKeys(ca.cluster.Name).Get(ca.cluster.ClusterConfig().Cloud.SSHKeyName)
```
Comment (tahsinrahman) [inline cloud/clusterapi.go]: remove store(ctx)

File: cloud/clusterapi.go [hunk 17]
```diff
@@ -233,19 +231,19 @@ func (ca *ClusterApi) CreatePharmerSecret() error {
 	}
 
 	if err = CreateSecret(ca.kc, "pharmer-certificate", ca.namespace, map[string][]byte{
```
Comment (tahsinrahman) [inline cloud/clusterapi.go]: required only for do, linode and packets. we should not create these secrets for all other providers

File: cloud/clusterapi.go [hunk 18]
```diff
@@ -233,19 +231,19 @@ func (ca *ClusterApi) CreatePharmerSecret() error {
...
+		"ca.key":             cert.EncodePrivateKeyPEM(ca.pharmerCertificates.CACert.Key),
+		"front-proxy-ca.crt": cert.EncodeCertPEM(ca.pharmerCertificates.FrontProxyCACert.Cert),
+		"front-proxy-ca.key": cert.EncodePrivateKeyPEM(ca.pharmerCertificates.FrontProxyCACert.Key),
+		"sa.crt":             cert.EncodeCertPEM(ca.pharmerCertificates.ServiceAccountCert.Cert),
+		"sa.key":             cert.EncodePrivateKeyPEM(ca.pharmerCertificates.ServiceAccountCert.Key),
 	}); err != nil {
 		return err
 	}
 
 	if err = CreateSecret(ca.kc, "pharmer-etcd", ca.namespace, map[string][]byte{
```
Comment (tahsinrahman) [inline cloud/clusterapi.go]: required only for do, linode and packets. we should not create these secrets for all other providers

File: cloud/cmds/options/apply_options.go [hunk 19]
```diff
@@ -30,6 +31,7 @@ func (c *ApplyConfig) AddFlags(fs *pflag.FlagSet) {
 }
 
 func (c *ApplyConfig) ValidateFlags(cmd *cobra.Command, args []string) error {
+	fmt.Println(args)
```
Comment (tahsinrahman) [inline cloud/cmds/options/apply_options.go]: remove

**Pull request merged**